Batch Operate Rollup table in Doris #2671

wangbo opened the pull request:
#2671

#2822

wangbo committed:
support batch add rollup

wangbo committed:
Merge branch 'upstream_master' into batch_alter_rollup_pr

wangbo committed:
fix code style

wangbo committed:
Batch Operate Rollup table in Doris #2671

wangbo committed:
fix ut error

morningman commented:
Plz wait for my review,  it will take some times.

I have a question:

For now, we support creating rollup for duplicate key table, not only aggregate key table. And we 
support syntex like:

`create materialized view as select a, b, sum(c) from tbl1 group by a, b;`

How to support this in your new `create table stmt`?

wangbo commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -2276,14 +2328,24 @@ cancel_stmt ::=
     :}
     ;
 
+cancel_rollup_job_id_list ::=
```
When user want to cancel job,he must first to execute "show alter table rollup" to get job info and find the job to be cancel,then use job info(jobid or rollup name) to cancel a job;
In my opinion,use rollup name isn't better than using jobid;In contrast, rollup name may be too long to cause alter sql too long.
Do you have some good reason to use rollup name?

wangbo commented:
> Plz wait for my review, it will take some times.
> 
> I have a question:
> 
> For now, we support creating rollup for duplicate key table, not only aggregate key table. And we
> support syntex like:
> 
> `create materialized view as select a, b, sum(c) from tbl1 group by a, b;`
> 
> How to support this in your new `create table stmt`?

1.  duplicate key table is olap table,so it is supported;
2.  I see "The materialized view is coming soon" in Catalog.createMaterializedView,create  MaterializedView is  not supported now ,so I didn't consider this case;

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -858,4 +1160,57 @@ public void cancel(CancelStmt stmt) throws DdlException {
+    // There may be dependency relation between user batch input rollup index
+    // e.g, user input rollup index: r1 from r2,r2 from r3 ;the input order must be r3,r2,r1
+    // sort rollup index for two reasons:
+    // 1. successfully create rollup index
```
This comment confuse me.
`sort rollup index` -> `successfully create rollup index` ?

wangbo replied:
Actually, rollup index can't be created in catalog before its parent rollup created,so does rollup job runs;so we must first sort rollup index,then  create index work and run rollup job;
I'll change the comment for better understand

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -85,6 +90,71 @@ public MaterializedViewHandler() {
+    // keep table's running job,used for concurrency limit
+    private Map<Long, Set<Long>> tableRunningJobMap = new ConcurrentHashMap<>();
+    // make sure that job's dependency just init once
+    private volatile boolean hasBuildJobDependency = false;
```
For simple. I think we needn't introduce the `Dependency` concept. We could check the
`alter statement` and require the from index must exist firstly, and then, we needn't consider the rollup job dependency.

wangbo replied:
Good suggestion;It seems check whether baseRollup can be found in catalog is enough.

wangbo replied:
No,Dependency is necessary;
Because int both parent job is cancel or parent is not final case, we all can't find parent index in catalog.Only 'parent index can be seen in catalog' can't decide whether current rollup job is  ready to run

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -590,15 +767,130 @@ protected void runAfterCatalogReady() {
+
+    private void changeTableStatus(long dbId, long tableId, OlapTableState olapTableState) {
+        Database db = Catalog.getCurrentCatalog().getDb(dbId);
+        db.readLock();
```
only use one write lock is enough.

wangbo replied:
Write lock means lower concurrency than read lock;

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -590,15 +767,130 @@ protected void runAfterCatalogReady() {
+        return true;
+    }
+
+    private void runAlterJobWithConcurrencyLimit(RollupJobV2 rollupJobV2) {
```
Would better add a comment to explain why concurrent rollup job for one table could work.

wangbo replied:
👌

kangkaisen commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -663,8 +665,54 @@ opt_user ::=
     :}
     ;
 
+add_rollup_clause ::=
```
Please update the user doc firstly.

wangbo replied:
👌

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -590,15 +767,130 @@ protected void runAfterCatalogReady() {
-            Map.Entry<Long, AlterJobV2> entry = iter.next();
-            AlterJobV2 alterJob = entry.getValue();
+        replayJobDependencyAfterCatalogReady();
+        Iterator<Map.Entry<Long, AlterJobV2>> iterator = getAlterJobsCopy().entrySet().iterator();
```
I think there is no concurrent issue. you could directly use `alterJobsV2`.

wangbo replied:
Alter Stmt write job to alterJobsV2，MaterializedViewHandler read the alterJobsV2;
So this is concurrent access;
To avoid unnecessary concurrent issue,I think better to read a alter job snapshot;

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -590,15 +767,130 @@ protected void runAfterCatalogReady() {
-        while (iter.hasNext()) {
-            Map.Entry<Long, AlterJobV2> entry = iter.next();
-            AlterJobV2 alterJob = entry.getValue();
+        replayJobDependencyAfterCatalogReady();
```
why replay here？

wangbo replied:
In this implementation,when fe restart,the job dependncy is lost,need to rebuild before running, set every job's parent job;
But if use "whether base rollup can be found in catalog" to judge whether parent rollup is ready is ok,so this method can be removed.

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -85,6 +90,71 @@ public MaterializedViewHandler() {
+        tableNotFinalStateJobIdSet.add(jobId);
+    }
+
+    private int removeAlterJobV2FromTableNotFinalStateJobMap(AlterJobV2 alterJobV2) {
```
change return type to bool?

wangbo replied:
👌

wangbo commented on fe/src/main/java/org/apache/doris/common/Config.java:
```diff
@@ -984,5 +984,11 @@
+     * control rollup job concurrent limit
+     */
+    @ConfField
+    public static int max_running_rollup_job_num = 1;
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/analysis/CreateTableStmt.java:
```diff
@@ -74,6 +75,7 @@
     private Map<String, String> extProperties;
     private String engineName;
     private String comment;
+    private List<AlterClause> ops;
```
or may be rollupOpts is better

wangbo commented on fe/src/main/java/org/apache/doris/analysis/CancelAlterTableStmt.java:
```diff
@@ -71,6 +85,11 @@ public String toSql() {
         StringBuilder stringBuilder = new StringBuilder();
         stringBuilder.append("CANCEL ALTER " + this.alterType);
         stringBuilder.append(" FROM " + dbTableName.toSql());
+        if (alterJobIdList != null && alterJobIdList.size() > 0) {
```
👌

wangbo committed:
1 update doc
2 persist batch rollup job in one atomic operator
3 init rollup job index when running

wangbo committed:
1 update doc
2 persist batch rollup job in one atomic operator
3 init rollup job index when running

wangbo committed:
Merge branch 'upstream/master' into batch_alter_rollup_pr_edit

wangbo committed:
Merge branch 'batch_alter_rollup_pr_edit' into batch_alter_rollup_pr

wangbo committed:
for github check

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -858,4 +1084,56 @@ public void cancel(CancelStmt stmt) throws DdlException {
+    // e.g, user input rollup index: r1 from r2,r2 from r3 ;the input order must be r3,r2,r1
+    // the reason for sort rollup index:
+    //   son rollup must be dealed after parent rollup, because they need parent rollup's info,such as baseRollupIndex
+    public static List<AlterClause> sortRollupIndex(List<AlterClause> alterClauseList) {
```
This is unnecessary. We could check the
`alter statement `and **require the from index must exist firstly**, and then, we needn't do this work

wangbo replied:
👌

kangkaisen commented:
> No,Dependency is necessary;
Because int both parent job is cancel or parent is not final case, we all can't find parent index in catalog.Only 'parent index can be seen in catalog' can't decide whether current rollup job is ready to run

I mean we don't need support  `user input rollup index: r1 from r2,r2 from r3`. **we could require the from index must exist firstly**. And then, there will no dependency between the rollup index, all children jobs all has the same parent job.

wangbo commented on fe/src/main/java/org/apache/doris/catalog/TabletInvertedIndex.java:
```diff
@@ -441,10 +441,10 @@ public void addTablet(long tabletId, TabletMeta tabletMeta) {
             if (!tabletMetaTable.contains(tabletMeta.getPartitionId(), tabletMeta.getIndexId())) {
                 tabletMetaTable.put(tabletMeta.getPartitionId(), tabletMeta.getIndexId(), tabletMeta);
-                LOG.debug("add tablet meta: {}", tabletId);
+                LOG.info("add tablet meta: {}", tabletId);
```
just for debug,it's useless

wangbo commented on fe/src/main/java/org/apache/doris/alter/BatchAlterJobV2.java:
```diff
@@ -0,0 +1,60 @@
+ * used for batch log AlterJob to editLog in one atomic operation
+ *
+ */
+public class BatchAlterJobV2 implements Writable {
```
👌

wangbo committed:
pr reivew 0203
1 remove dependency bewteen rollupjob
2 add ut

wangbo commented:
@morningman plz reivew this pr

kangkaisen commented on docs/documentation/cn/sql-reference/sql-statements/Data Definition/CANCEL ALTER.md:
```diff
@@ -29,7 +29,11 @@ under the License.
         FROM db_name.table_name
-        
+
+    3. 根据job id批量撤销rollup操作
```
英文版?

wangbo replied:
这个是documentation/cn啊，应该就是中文

wangbo replied:
👌

kangkaisen commented on fe/src/main/java/org/apache/doris/catalog/Catalog.java:
```diff
@@ -3543,6 +3548,32 @@ private void createOlapTable(Database db, CreateTableStmt stmt) throws DdlExcept
+        for (AlterClause alterClause : stmt.getRollupAlterClauseList()) {
+            AddRollupClause addRollupClause = (AddRollupClause)alterClause;
+
+            // TODO(wangbo): 2020/2/3 doris auto select base rollup
```
Remove this comment.  One Olap table always has one base index.

wangbo replied:
👌

kangkaisen commented on fe/src/main/java/org/apache/doris/alter/RollupJobV2.java:
```diff
@@ -305,7 +305,9 @@ protected void runWaitingTxnJob() throws AlterCancelException {
         }
-        
+
+        // TODO(wangbo): 2020/2/4 in the case that the [WaitingTxnJob|PendingJob] is checkpointed and replayMethod won't be called, make sure the tablet meta exists in olapTable and TabletInvertedIndex before doris run
```
what's the meaning of this comment?
After checkpoint, we must replay the journal less than checkPointVersion

wangbo replied:
TabletInvertedIndex and MaterializedIndex in olaptable can't be persisted until job is final,they     
depend on replay to load data to memory

morningman replied:
I think this comment can be removed when you keep the job's state
in PENDING after calling `replayPending()`. Am I right?

kangkaisen commented:
Please add doc for create table with rollup

wangbo commented on fe/src/main/java/org/apache/doris/alter/AlterJobV2.java:
```diff
@@ -38,7 +38,8 @@
 
     public enum JobState {
-        PENDING, // Job is created
+        INIT, // Job is created
```
not necessary now

wangbo commented on fe/src/main/java/org/apache/doris/alter/BatchAlterJobV2.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
```
RollupJobV2 not support json serde,so BatchAlterJobV2 can't Serde by JSON

wangbo commented on docs/documentation/cn/sql-reference/sql-statements/Data Definition/ALTER TABLE.md:
```diff
@@ -63,20 +63,19 @@ under the License.
         
     rollup 支持如下几种创建方式：
-    1. 创建 rollup index
+    1. 批量创建 rollup index
```
👌

wangbo committed:
1 supplement doc
2 remove useless comment

morningman commented on fe/src/main/java/org/apache/doris/alter/RollupJobV2.java:
```diff
@@ -614,33 +614,35 @@ private void replayPending(RollupJobV2 replayedJob) {
         }
 
-        this.jobState = JobState.WAITING_TXN;
+        this.jobState = JobState.PENDING;
```
Add a comment here to explain why after replaying the pending job,
the state is still PENDING:

`to make sure that this job will run runPendingJob() again to create the rollup replicas`

wangbo replied:
👌

morningman commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -507,6 +608,12 @@ private long checkAndGetBaseIndex(String baseIndexName, OlapTable olapTable) thr
 
+    public void processBatchDropRollup (List<AlterClause> dropRollupClauses, Database db, OlapTable olapTable) throws DdlException {
+        for (AlterClause dropRollupClause : dropRollupClauses) {
+            processDropRollup((DropRollupClause)dropRollupClause, db, olapTable);
```
"batch drop rollup" operation should still be atomic when writing meta log.
So you need to modify the `processDropRollup()`:

1. First, traverse all drop clauses to validate them, such as whether to drop the base index, does the rollup index need to be dropped exist.

2. Do the real drop operation. And all-or-non rollups should be dropped.

3. Write a edit log named `BatchDropRollupOperationLog` to persist the meta.

morningman commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -85,6 +88,54 @@ public MaterializedViewHandler() {
 
+
+    // for batch submit rollup job, tableId -> jobId
+    // keep table's not final state job size. The job size determine's table's state, = 0 means table is normal,otherwrise is rollup
```
```suggestion
    // keep table's not final state job size. The job size determine's table's state, = 0 means table is normal, otherwise is rollup
```

wangbo replied:
👌

morningman commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -85,6 +88,54 @@ public MaterializedViewHandler() {
+
+    private void addAlterJobV2ToTableNotFinalStateJobMap(AlterJobV2 alterJobV2) {
+        if (alterJobV2.isDone()) {
+            LOG.warn("try to add a final job({}) to a unfinal set", alterJobV2.getJobId());
```
Can it be an assertion? I found there is no way to add an already-done job here.

wangbo replied:
prevent later change will add an already-done job;It seems assertion is better

wangbo replied:
job replay will add a done job

morningman commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -85,6 +88,54 @@ public MaterializedViewHandler() {
+        tableNotFinalStateJobIdSet.add(jobId);
+    }
+
+    private boolean removeAlterJobV2FromTableNotFinalStateJobMap(AlterJobV2 alterJobV2) {
```
Add comment to explain return value's meaning.

wangbo replied:
👌

wangbo commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -787,11 +1079,15 @@ private void getOldAlterJobInfos(Database db, List<List<Comparable>> rollupJobIn
-        for (AlterClause alterClause : alterClauses) {
+        alterClauses = sortRollupIndex(alterClauses);
+
+        Iterator<AlterClause> alterClauseIterator = alterClauses.iterator();
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/analysis/CreateTableStmt.java:
```diff
@@ -218,6 +236,14 @@ public String getComment() {
         return comment;
     }
 
+    public void setRollupAlterClauseList(List<AlterClause> rollupAlterClauseList) {
```
👌

wangbo commented on docs/documentation/cn/sql-reference/sql-statements/Data Definition/CREATE TABLE.md:
```diff
@@ -550,6 +559,26 @@ under the License.
+    rollup (
+    r1(event_day,siteid),
+    r2(event_day,citycode),
+    r3(event_day) from r1
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/catalog/Catalog.java:
```diff
@@ -1682,7 +1683,11 @@ public long loadAlterJob(DataInputStream dis, long checksum, JobType type) throw
             for (int i = 0; i < size; i++) {
                 AlterJobV2 alterJobV2 = AlterJobV2.read(dis);
-                alterJobsV2.put(alterJobV2.getJobId(), alterJobV2);
+                if (type == JobType.ROLLUP) {
```
No bug before; Before,SchemaChangeHandler has the `addJob` logic with MaterializedViewHandler;But now,they are different;

wangbo commented on fe/src/main/java/org/apache/doris/common/Config.java:
```diff
@@ -984,5 +984,11 @@
+    /*
+     * control rollup job concurrent limit
+     */
+    @ConfField
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/persist/OperationType.java:
```diff
@@ -56,6 +56,7 @@
     public static final short OP_FINISH_CONSISTENCY_CHECK = 29;
     public static final short OP_RENAME_ROLLUP = 120;
     public static final short OP_ALTER_JOB_V2 = 121;
+    public static final short OP_BATCH_ALTER_JOB_V2 = 123;
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java:
```diff
@@ -1436,7 +1437,8 @@ public void cancel(CancelStmt stmt) throws DdlException {
             // find from new alter jobs first
-            schemaChangeJobV2 = getUnfinishedAlterJobV2(olapTable.getId());
+            List<AlterJobV2> schemaChangeJobV2List = getUnfinishedAlterJobV2ByTableId(olapTable.getId());
+            schemaChangeJobV2 = schemaChangeJobV2List.size() == 0 ? schemaChangeJobV2 : Iterables.getOnlyElement(schemaChangeJobV2List);
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/alter/RollupJobV2.java:
```diff
@@ -305,7 +305,9 @@ protected void runWaitingTxnJob() throws AlterCancelException {
         }
-        
+
+        // TODO(wangbo): 2020/2/4 in the case that the [WaitingTxnJob|PendingJob] is checkpointed and replayMethod won't be called, make sure the tablet meta exists in olapTable and TabletInvertedIndex before doris run
```
👌

wangbo commented on fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:
```diff
@@ -830,8 +1008,14 @@ public void cancel(CancelStmt stmt) throws DdlException {
-            if (rollupJobV2 == null) {
+            if (cancelAlterTableStmt.getAlterJobIdList() != null) {
+                for (Long jobId : cancelAlterTableStmt.getAlterJobIdList()) {
+                    rollupJobV2List.add(getUnfinishedAlterJobV2ByJobId(jobId));
```
In current implementation,all rollup job is independent,`partial success` just affect user experience,atomic batch operation is not necessary.
I prefer execute batch operation in async way, user submit a stmt, and use another stmt to check result.

wangbo replied:
👌

wangbo committed:
0208 pr review

wangbo committed:
2010 review
use json to serialize BatchDropInfo

morningman reviewed (approved):
LGTM
